File: java/zebra-puzzle/src/main/java/AxiomStatement.java
final class AxiomStatement implements Statement {
    private final int houseIdx;
    private final String param;
    private final String value;

    AxiomStatement(int houseIdx, String param, String value) {
        this.houseIdx = houseIdx;
        this.param = param;
        this.value = value;
    }

    @Override
    public ApplyResult apply(State state) {
        state.set(state.houses.get(houseIdx), param, value);
        return ApplyResult.DONE;
    }

    @Override
    public Compatibility getCompatibility(State state) {
        return value.equals(state.houses.get(houseIdx).get(param))
                ? Compatibility.PERFECT_MATCH
                : Compatibility.INCOMPATIBLE;
    }
}

File: kotlin/zebra-puzzle/_iterations/1/src/main/kotlin/AxiomStatement.kt
class AxiomStatement(private val houseIdx: Int, private val param: String, private val value: String) : Statement {
    override fun apply(state: State): ApplyResult {
        state.set(state.houses[houseIdx], param, value)
        return ApplyResult.DONE
    }

    override fun getCompatibility(state: State): Compatibility = when (value) {
        state.houses[houseIdx].get(param) -> Compatibility.PERFECT_MATCH
        else -> Compatibility.INCOMPATIBLE
    }
}
